common/types/overflow.go

```diff
@@ -222,57 +259,99 @@ func subtractTimeChecked(x, y time.Time) (time.Duration, bool) {
-			nsec += int64(time.Second)
-		}
-	}
-
 	// Scale seconds to nanoseconds detecting overflow.
-	tsec, ok := multiplyInt64Checked(sec, int64(time.Second))
-	if !ok {
-		return time.Duration(0), false
+	tsec, err := multiplyInt64Checked(sec, int64(time.Second))
+	if err != nil {
+		return time.Duration(0), err
 	}
 
 	// Lastly we need to add the two nanoseconds together.
-	val, ok := addInt64Checked(tsec, nsec)
-	if !ok {
-		return time.Duration(0), false
+	val, err := addInt64Checked(tsec, nsec)
+	if err != nil {
+		return time.Duration(0), err
 	}
 
-	return time.Duration(val), true
+	return time.Duration(val), nil
 }
 
 // subtractTimeDurationChecked performs subtraction with overflow detection of a time.Time and
-// time.Duration, returning the result of the subtraction if no overflow occurred as the first
-// return value and a bool indicating whether no overflow occurred as the second return value.
-func subtractTimeDurationChecked(x time.Time, y time.Duration) (time.Time, bool) {
+// time.Duration.
+//
+// If the operation fails due to overflow the error return value will be non-nil.
+func subtractTimeDurationChecked(x time.Time, y time.Duration) (time.Time, error) {
 	// The easiest way to implement this is to negate y and add them.
 	// x - y = x + -y
-	val, ok := negateDurationChecked(y)
-	if !ok {
-		return time.Time{}, false
+	val, err := negateDurationChecked(y)
+	if err != nil {
+		return time.Time{}, err
 	}
 	return addTimeDurationChecked(x, val)
 }
+
+// doubleToInt64Checked converts a double to an int64 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func doubleToInt64Checked(v float64) (int64, error) {
+	if math.IsInf(v, 0) || v <= float64(math.MinInt64) || v >= float64(math.MaxInt64) {
+		return 0, errIntOverflow
+	}
+	return int64(v), nil
+}
+
+// doubleToInt64Checked converts a double to a uint64 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func doubleToUint64Checked(v float64) (uint64, error) {
+	if math.IsInf(v, 0) || v < 0 || v >= doubleTwoTo64 {
+		return 0, errUintOverflow
+	}
+	return uint64(v), nil
+}
+
+// int64toUint64Checked converts an int64 to a uint64 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func int64ToUint64Checked(v int64) (uint64, error) {
+	if v < 0 {
+		return 0, errUintOverflow
+	}
+	return uint64(v), nil
+}
+
+// int64toInt32Checked converts an int64 to an int32 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func int64ToInt32Checked(v int64) (int32, error) {
+	if v < math.MinInt32 || v > math.MaxInt32 {
+		return 0, errIntOverflow
+	}
+	return int32(v), nil
+}
+
+// uint64toUint32Checked converts a uint64 to a uint32 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func uint64ToUint32Checked(v uint64) (uint32, error) {
+	if v < 0 || v > math.MaxUint32 {
```

Comment:
v is unsigned, so it is never less than 0.

Reply:
Yeah, not sure why I felt the need to add that. Removed.